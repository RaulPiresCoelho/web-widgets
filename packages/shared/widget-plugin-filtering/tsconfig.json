{
    "extends": "@mendix/tsconfig-web-widgets/esm-library-with-jsx",
    "include": ["./src/**/*"],
    "compilerOptions": {
        "outDir": "./dist",
<<<<<<< HEAD
        "rootDir": "./src",
        "resolveJsonModule": true
=======
        "rootDir": "./src"
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    }
}
