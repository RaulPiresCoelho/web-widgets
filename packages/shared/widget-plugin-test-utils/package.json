{
  "name": "@mendix/widget-plugin-test-utils",
  "version": "0.1.0",
  "description": "Test utils for pluggable widgets.",
  "license": "Apache-2.0",
  "copyright": "Â© Mendix Technology BV 2023. All rights reserved.",
  "repository": {
    "type": "git",
    "url": "https://github.com/mendix/web-widgets.git"
  },
  "type": "module",
  "main": "./dist/main.js",
  "exports": {
<<<<<<< HEAD
    ".": "./dist/main.js",
    "./__mocks__/*": "./__mocks__/*"
=======
    ".": "./dist/main.js"
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
  },
  "typesVersions": {
    "*": {
      "*": [
        "./dist/main.d.ts"
      ]
    }
  },
  "scripts": {
<<<<<<< HEAD
    "prepare": "pnpm build",
    "lint": "eslint --ext .ts,.tsx ./src",
    "format": "prettier --write .",
    "build": "tsc -p tsconfig.build.json",
    "dev": "tsc -p tsconfig.build.json --watch",
    "test": "jest"
=======
    "prepare": "tsc",
    "lint": "eslint --ext .ts,.tsx ./src",
    "format": "prettier --write .",
    "dev": "tsc --watch",
    "compile": "tsc"
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
  },
  "devDependencies": {
    "@mendix/eslint-config-web-widgets": "workspace:*",
    "@mendix/prettier-config-web-widgets": "workspace:*",
<<<<<<< HEAD
    "@mendix/tsconfig-web-widgets": "workspace:*",
    "@swc/core": "^1.7.26",
    "@swc/jest": "^0.2.36",
    "big.js": "^6.2.2"
  },
  "dependencies": {
    "mendix": "^10.16.49747"
=======
    "@mendix/tsconfig-web-widgets": "workspace:*"
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
  }
}
