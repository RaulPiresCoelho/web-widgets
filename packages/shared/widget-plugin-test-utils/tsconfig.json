{
<<<<<<< HEAD
    "extends": "./tsconfig.build.json",
    "include": ["./node_modules/mendix/index.d.ts", "./src/**/*", "./__tests__/**/*", "__mocks__/**/*"],
    "compilerOptions": {
        "noEmit": true,
        "rootDir": "."
=======
    "extends": "@mendix/tsconfig-web-widgets/esm-library-with-jsx",
    "include": ["./src/**/*"],
    "compilerOptions": {
        "outDir": "./dist",
        "rootDir": "./src"
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    }
}
