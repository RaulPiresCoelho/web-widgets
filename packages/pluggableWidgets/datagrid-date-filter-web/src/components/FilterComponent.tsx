<<<<<<< HEAD
import { FilterSelector } from "@mendix/widget-plugin-filtering/controls";
import classNames from "classnames";
import { createElement, ReactElement } from "react";
import { FilterTypeEnum } from "../helpers/base-types";
import { DatePicker, DatePickerProps } from "./DatePicker";

export interface FilterComponentProps extends DatePickerProps {
    id?: string;
    class: string;
    tabIndex?: number;
    style?: React.CSSProperties;
    placeholder?: string;
    screenReaderButtonCaption?: string;
    screenReaderCalendarCaption?: string;
    screenReaderInputCaption?: string;
    filterFn?: FilterTypeEnum;
    onFilterChange: (fn: FilterTypeEnum) => void;
=======
import { FilterSelector } from "@mendix/widget-plugin-filter-selector/FilterSelector";
import classNames from "classnames";
import { createElement, ReactElement } from "react";
import { FilterTypeEnum } from "../helpers/base-types";
import { DatePicker } from "./DatePicker";
import { SetupProps, useSetup } from "../helpers/useSetup";
import { useReset } from "../helpers/useReset";

export interface FilterComponentProps extends SetupProps {
    name: string;
    adjustable: boolean;
    class: string;
    tabIndex: number;
    defaultFilter: FilterTypeEnum;
    style?: React.CSSProperties;
    placeholder?: string;
    parentChannelName?: string | null;
    screenReaderButtonCaption?: string;
    screenReaderCalendarCaption?: string;
    screenReaderInputCaption?: string;
    defaultValue?: Date;
    defaultStartDate?: Date;
    defaultEndDate?: Date;
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
}

export type FilterComponent = typeof FilterComponent;

export function FilterComponent(props: FilterComponentProps): ReactElement {
<<<<<<< HEAD
=======
    const { id, filterStore, calendarStore, datePickerController } = useSetup(props);
    useReset(props, filterStore);
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    return (
        <div
            className={classNames("filter-container", props.class)}
            data-focusindex={props.tabIndex}
            style={props.style}
        >
            {props.adjustable && (
                <FilterSelector
                    ariaLabel={props.screenReaderButtonCaption ?? "Select filter type"}
<<<<<<< HEAD
                    value={props.filterFn ?? "equal"}
                    onChange={props.onFilterChange}
                    options={OPTIONS}
                />
            )}
            <DatePicker {...props} />
=======
                    defaultFilter={filterStore.state.filterType}
                    id={id}
                    onChange={filterStore.setType}
                    options={OPTIONS}
                />
            )}
            <DatePicker
                adjustable={props.adjustable}
                parentId={id}
                placeholder={props.placeholder}
                screenReaderCalendarCaption={props.screenReaderCalendarCaption}
                screenReaderInputCaption={props.screenReaderInputCaption}
                filterStore={filterStore}
                calendarStore={calendarStore}
                datePickerController={datePickerController}
            />
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
        </div>
    );
}

const OPTIONS = [
    { value: "between", label: "Between" },
    { value: "greater", label: "Greater than" },
    { value: "greaterEqual", label: "Greater than or equal" },
    { value: "equal", label: "Equal" },
    { value: "notEqual", label: "Not equal" },
    { value: "smaller", label: "Smaller than" },
    { value: "smallerEqual", label: "Smaller than or equal" },
    { value: "empty", label: "Empty" },
    { value: "notEmpty", label: "Not empty" }
] as Array<{ value: FilterTypeEnum; label: string }>;
