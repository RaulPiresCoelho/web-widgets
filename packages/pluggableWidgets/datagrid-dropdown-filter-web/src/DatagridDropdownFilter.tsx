<<<<<<< HEAD
import { createElement, ReactElement } from "react";
import { withPreloader } from "@mendix/widget-plugin-platform/hoc/withPreloader";
import { DatagridDropdownFilterContainerProps } from "../typings/DatagridDropdownFilterProps";
import { StaticFilterContainer } from "./components/StaticFilterContainer";
import { withSelectFilterAPI, Select_FilterAPIv2 } from "./hocs/withSelectFilterAPI";
import { RefFilterContainer } from "./components/RefFilterContainer";

function Container(props: DatagridDropdownFilterContainerProps & Select_FilterAPIv2): React.ReactElement {
    const commonProps = {
        ariaLabel: props.ariaLabel?.value,
        className: props.class,
        tabIndex: props.tabIndex,
        styles: props.style,
        onChange: props.onChange,
        valueAttribute: props.valueAttribute,
        parentChannelName: props.parentChannelName,
        name: props.name
    };

    if (props.filterStore.type === "refselect") {
        return (
            <RefFilterContainer
                {...commonProps}
                filterStore={props.filterStore}
                multiselect={props.multiSelect}
                emptyCaption={props.emptyOptionCaption?.value}
            />
        );
    }

    return (
        <StaticFilterContainer
            {...commonProps}
            filterStore={props.filterStore}
            filterOptions={props.filterOptions}
            multiselect={props.multiSelect}
            defaultValue={props.defaultValue?.value}
            emptyCaption={props.emptyOptionCaption?.value}
        />
    );
}

const container = withPreloader(Container, props => props.defaultValue?.status === "loading");

const Widget = withSelectFilterAPI(container);

export default function DatagridDropdownFilter(props: DatagridDropdownFilterContainerProps): ReactElement {
    return <Widget {...props} />;
=======
import { useFilterContextValue } from "@mendix/widget-plugin-filtering";
import { createElement, Fragment, ReactElement } from "react";
import { DatagridDropdownFilterContainerProps } from "../typings/DatagridDropdownFilterProps";
import { EnumerationFilter } from "./components/EnumerationFilter";
import { ErrorBox } from "./components/ErrorBox";
import { AssociationFilter } from "./components/AssociationFilter";
import { OutOfProviderError } from "./features/errors";

export default function DatagridDropdownFilter(props: DatagridDropdownFilterContainerProps): ReactElement {
    const context = useFilterContextValue();

    if (context.hasError) {
        return <ErrorBox error={new OutOfProviderError()} />;
    }

    if (props.defaultValue?.status === "loading") {
        return <Fragment />;
    }

    const Filter = context.value.associationProperties ? AssociationFilter : EnumerationFilter;

    return <Filter context={context.value} widgetProps={props} />;
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
}
