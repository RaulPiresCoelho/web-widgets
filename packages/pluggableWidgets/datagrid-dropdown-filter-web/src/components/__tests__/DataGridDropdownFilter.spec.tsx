import "@testing-library/jest-dom";
<<<<<<< HEAD
import { FilterAPIv2 } from "@mendix/widget-plugin-filtering/context";
import { HeaderFiltersStore, HeaderFiltersStoreProps } from "@mendix/widget-plugin-filtering/stores/HeaderFiltersStore";
import { dynamicValue, ListAttributeValueBuilder } from "@mendix/widget-plugin-test-utils";
import { createContext, createElement } from "react";
import DatagridDropdownFilter from "../../DatagridDropdownFilter";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";

export interface StaticInfo {
    name: string;
    filtersChannelName: string;
}
=======
import { Alert } from "@mendix/widget-plugin-component-kit/Alert";
import { FilterContextValue } from "@mendix/widget-plugin-filtering";
import {
    actionValue,
    dynamicValue,
    EditableValueBuilder,
    ListAttributeValueBuilder
} from "@mendix/widget-plugin-test-utils";
import { createContext, createElement } from "react";
import DatagridDropdownFilter from "../../DatagridDropdownFilter";
import userEvent from "@testing-library/user-event";
import { render, screen, waitFor } from "@testing-library/react";
import { mount } from "enzyme";
import { FilterComponent } from "../FilterComponent";
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)

const commonProps = {
    class: "filter-custom-class",
    tabIndex: 0,
    name: "filter-test",
<<<<<<< HEAD
    advanced: false,
    groupKey: "dropdown-filter"
};

const headerFilterStoreInfo: StaticInfo = {
    name: commonProps.name,
    filtersChannelName: ""
=======
    advanced: false
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
};

describe("Dropdown Filter", () => {
    describe("with single instance", () => {
        afterEach(() => {
            delete (global as any)["com.mendix.widgets.web.UUID"];
        });

        describe("with single attribute", () => {
            function mockCtx(universe: string[]): void {
<<<<<<< HEAD
                const props: HeaderFiltersStoreProps = {
                    filterList: [
                        {
                            filter: new ListAttributeValueBuilder()
                                .withUniverse(universe)
                                .withType("Enum")
                                .withFilterable(true)
                                .withFormatter(
                                    value => value,
                                    () => console.log("Parsed")
                                )
                                .build()
                        }
                    ]
                };
                const headerFilterStore = new HeaderFiltersStore(props, headerFilterStoreInfo, null);
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = createContext<FilterAPIv2>(
                    headerFilterStore.context
                );
            }
            beforeEach(() => {
=======
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                    filterDispatcher: jest.fn(),
                    singleAttribute: new ListAttributeValueBuilder()
                        .withUniverse(universe)
                        .withType("Enum")
                        .withFilterable(true)
                        .withFormatter(
                            value => value,
                            () => console.log("Parsed")
                        )
                        .build()
                } as FilterContextValue);
            }
            beforeAll(() => {
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
                mockCtx(["enum_value_1", "enum_value_2"]);
            });

            describe("with auto options", () => {
<<<<<<< HEAD
                it("loads correct values from universe", async () => {
                    const filter = render(
                        <DatagridDropdownFilter {...commonProps} auto multiSelect={false} filterOptions={[]} />
                    );

                    const trigger = filter.getByRole("textbox");

                    await fireEvent.click(trigger);

                    const items = filter.getAllByRole("menuitem");

                    items.forEach((item, index) => {
                        if (index === 0) {
                            return;
                        }
                        expect(item.textContent).toEqual(`enum_value_${index}`);
                    });
=======
                it("loads correct values from universe", () => {
                    const filter = mount(
                        <DatagridDropdownFilter {...commonProps} auto multiSelect={false} filterOptions={[]} />
                    );

                    expect(filter.find(FilterComponent).prop("options")).toStrictEqual([
                        {
                            caption: "enum_value_1",
                            value: "enum_value_1"
                        },
                        {
                            caption: "enum_value_2",
                            value: "enum_value_2"
                        }
                    ]);
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
                });
            });

            describe("DOM structure", () => {
                it("renders correctly", () => {
                    const { asFragment } = render(
                        <DatagridDropdownFilter {...commonProps} auto multiSelect={false} filterOptions={[]} />
                    );

                    expect(asFragment()).toMatchSnapshot();
                });
            });

<<<<<<< HEAD
=======
            it("triggers attribute and onchange action on change filter value", async () => {
                const action = actionValue();
                const attribute = new EditableValueBuilder<string>().build();
                render(
                    <DatagridDropdownFilter
                        {...commonProps}
                        auto
                        multiSelect={false}
                        filterOptions={[]}
                        onChange={action}
                        valueAttribute={attribute}
                    />
                );

                await userEvent.click(screen.getByRole("textbox"));
                await userEvent.click(screen.getAllByRole("menuitem")[2]);

                expect(action.execute).toBeCalledTimes(1);
                expect(attribute.setValue).toBeCalledWith("enum_value_2");
            });

>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            describe("with defaultValue", () => {
                it("initialize component with defaultValue", () => {
                    render(
                        <DatagridDropdownFilter
                            {...commonProps}
                            auto
                            multiSelect={false}
                            filterOptions={[]}
                            defaultValue={dynamicValue<string>("enum_value_1")}
                        />
                    );

                    expect(screen.getByRole("textbox")).toHaveValue("enum_value_1");
                });

                it("don't sync defaultValue with state when defaultValue changes from undefined to string", async () => {
                    const { rerender } = render(
                        <DatagridDropdownFilter
                            {...commonProps}
                            auto
                            multiSelect={false}
                            filterOptions={[]}
                            defaultValue={dynamicValue<string>("")}
                        />
                    );

                    await waitFor(() => {
                        expect(screen.getByRole("textbox")).toHaveValue("");
                    });

                    // “Real” context causes widgets to re-renders multiple times, replicate this in mocked context.
                    rerender(
                        <DatagridDropdownFilter
                            {...commonProps}
                            auto
                            multiSelect={false}
                            filterOptions={[]}
                            defaultValue={dynamicValue<string>("")}
                        />
                    );
                    rerender(
                        <DatagridDropdownFilter
                            {...commonProps}
                            auto
                            multiSelect={false}
                            filterOptions={[]}
                            defaultValue={dynamicValue<string>("enum_value_1")}
                        />
                    );

                    await waitFor(() => {
                        expect(screen.getByRole("textbox")).toHaveValue("");
                    });
                });

                it("don't sync defaultValue with state when defaultValue changes from string to undefined", async () => {
                    mockCtx(["xyz", "abc"]);
                    const { rerender } = render(
                        <DatagridDropdownFilter
                            {...commonProps}
                            auto
                            multiSelect={false}
                            filterOptions={[]}
                            defaultValue={dynamicValue<string>("xyz")}
                        />
                    );

                    expect(screen.getByRole("textbox")).toHaveValue("xyz");

                    // “Real” context causes widgets to re-renders multiple times, replicate this in mocked context.
                    rerender(
                        <DatagridDropdownFilter
                            {...commonProps}
                            auto
                            multiSelect={false}
                            filterOptions={[]}
                            defaultValue={dynamicValue<string>("xyz")}
                        />
                    );
                    rerender(
                        <DatagridDropdownFilter
                            {...commonProps}
                            auto
                            multiSelect={false}
                            filterOptions={[]}
                            defaultValue={undefined}
                        />
                    );

                    await waitFor(() => {
                        expect(screen.getByRole("textbox")).toHaveValue("xyz");
                    });
                });
            });

            afterAll(() => {
<<<<<<< HEAD
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = undefined;
=======
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            });
        });

        describe("with multiple attributes", () => {
            beforeAll(() => {
<<<<<<< HEAD
                const props: HeaderFiltersStoreProps = {
                    filterList: [
                        {
                            filter: new ListAttributeValueBuilder()
                                .withId("attribute1")
                                .withUniverse(["enum_value_1", "enum_value_2"])
                                .withType("Enum")
                                .withFilterable(true)
                                .withFormatter(
                                    value => value,
                                    () => console.log("Parsed")
                                )
                                .build()
                        },
                        {
                            filter: new ListAttributeValueBuilder()
                                .withId("attribute2")
                                .withUniverse([true, false])
                                .withType("Boolean")
                                .withFilterable(true)
                                .withFormatter(
                                    value => (value ? "Yes" : "No"),
                                    () => console.log("Parsed")
                                )
                                .build()
                        }
                    ]
                };
                const headerFilterStore = new HeaderFiltersStore(props, headerFilterStoreInfo, null);
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = createContext<FilterAPIv2>(
                    headerFilterStore.context
                );
            });

            describe("with auto options", () => {
                it("loads correct values from universes", async () => {
                    const filter = render(
                        <DatagridDropdownFilter {...commonProps} auto multiSelect={false} filterOptions={[]} />
                    );

                    const trigger = filter.getByRole("textbox");
                    await fireEvent.click(trigger);

                    expect(filter.getAllByRole("menuitem").map(item => item.textContent)).toStrictEqual([
                        "",
                        "enum_value_1",
                        "enum_value_2",
                        "Yes",
                        "No"
=======
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                    filterDispatcher: jest.fn(),
                    multipleAttributes: {
                        attribute1: new ListAttributeValueBuilder()
                            .withId("attribute1")
                            .withUniverse(["enum_value_1", "enum_value_2"])
                            .withType("Enum")
                            .withFilterable(true)
                            .withFormatter(
                                value => value,
                                () => console.log("Parsed")
                            )
                            .build(),
                        attribute2: new ListAttributeValueBuilder()
                            .withId("attribute2")
                            .withUniverse([true, false])
                            .withType("Boolean")
                            .withFilterable(true)
                            .withFormatter(
                                value => (value ? "Yes" : "No"),
                                () => console.log("Parsed")
                            )
                            .build()
                    }
                } as FilterContextValue);
            });

            describe("with auto options", () => {
                it("loads correct values from universes", () => {
                    const filter = mount(
                        <DatagridDropdownFilter {...commonProps} auto multiSelect={false} filterOptions={[]} />
                    );

                    expect(filter.find(FilterComponent).prop("options")).toStrictEqual([
                        {
                            caption: "enum_value_1",
                            value: "enum_value_1"
                        },
                        {
                            caption: "enum_value_2",
                            value: "enum_value_2"
                        },
                        {
                            caption: "Yes",
                            value: "true"
                        },
                        {
                            caption: "No",
                            value: "false"
                        }
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
                    ]);
                });
            });

            afterAll(() => {
<<<<<<< HEAD
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = undefined;
=======
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            });
        });

        describe("with wrong attribute's type", () => {
            beforeAll(() => {
<<<<<<< HEAD
                const props: HeaderFiltersStoreProps = {
                    filterList: [
                        {
                            filter: new ListAttributeValueBuilder().withType("String").withFilterable(true).build()
                        }
                    ]
                };
                const headerFilterStore = new HeaderFiltersStore(props, headerFilterStoreInfo, null);
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = createContext<FilterAPIv2>(
                    headerFilterStore.context
                );
            });

            it("renders error message", () => {
                const { container } = render(
                    <DatagridDropdownFilter {...commonProps} auto multiSelect={false} filterOptions={[]} />
                );

                expect(container.querySelector(".alert")?.textContent).toBe(
                    "Unable to get filter store. Check parent widget configuration."
=======
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                    filterDispatcher: jest.fn(),
                    singleAttribute: new ListAttributeValueBuilder().withType("String").withFilterable(true).build()
                } as FilterContextValue);
            });

            it("renders error message", () => {
                const filter = mount(
                    <DatagridDropdownFilter {...commonProps} auto multiSelect={false} filterOptions={[]} />
                );

                expect(filter.find(Alert).text()).toBe(
                    `Error: The attribute type being used for Drop-down filter is not 'Boolean or Enumeration'.`
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
                );
            });

            afterAll(() => {
<<<<<<< HEAD
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = undefined;
=======
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            });
        });

        describe("with wrong multiple attributes' types", () => {
            beforeAll(() => {
<<<<<<< HEAD
                const props: HeaderFiltersStoreProps = {
                    filterList: [
                        {
                            filter: new ListAttributeValueBuilder()
                                .withId("attribute1")
                                .withType("String")
                                .withFilterable(true)
                                .build()
                        },
                        {
                            filter: new ListAttributeValueBuilder()
                                .withId("attribute2")
                                .withType("Decimal")
                                .withFilterable(true)
                                .build()
                        }
                    ]
                };
                const headerFilterStore = new HeaderFiltersStore(props, headerFilterStoreInfo, null);
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = createContext<FilterAPIv2>(
                    headerFilterStore.context
                );
            });

            it("renders error message", () => {
                const { container } = render(
                    <DatagridDropdownFilter {...commonProps} auto multiSelect={false} filterOptions={[]} />
                );

                expect(container.querySelector(".alert")?.textContent).toBe(
                    "Unable to get filter store. Check parent widget configuration."
=======
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                    filterDispatcher: jest.fn(),
                    multipleAttributes: {
                        attribute1: new ListAttributeValueBuilder()
                            .withId("attribute1")
                            .withType("String")
                            .withFilterable(true)
                            .build(),
                        attribute2: new ListAttributeValueBuilder()
                            .withId("attribute2")
                            .withType("Decimal")
                            .withFilterable(true)
                            .build()
                    }
                } as FilterContextValue);
            });

            it("renders error message", () => {
                const filter = mount(
                    <DatagridDropdownFilter {...commonProps} auto multiSelect={false} filterOptions={[]} />
                );

                expect(filter.find(Alert).text()).toBe(
                    `Error: Missing required attribute(s): the Drop-down filter widget can't be used with the filters options you have selected. It requires a 'Boolean or Enumeration' attribute to be selected.`
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
                );
            });

            afterAll(() => {
<<<<<<< HEAD
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = undefined;
=======
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            });
        });

        describe("with no context", () => {
            beforeAll(() => {
<<<<<<< HEAD
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = undefined;
            });

            it("renders error message", () => {
                const { container } = render(
                    <DatagridDropdownFilter {...commonProps} auto multiSelect={false} filterOptions={[]} />
                );

                expect(container.querySelector(".alert")?.textContent).toBe(
                    "The filter widget must be placed inside the column or header of the Data grid 2.0 or inside header of the Gallery widget."
=======
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
            });

            it("renders error message", () => {
                const filter = mount(
                    <DatagridDropdownFilter {...commonProps} auto multiSelect={false} filterOptions={[]} />
                );

                expect(filter.find(Alert).text()).toBe(
                    "Error: The Drop-down filter widget must be placed inside the header of the Data grid 2.0 or Gallery widget."
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
                );
            });
        });

        describe("with invalid values", () => {
            beforeAll(() => {
<<<<<<< HEAD
                const props: HeaderFiltersStoreProps = {
                    filterList: [
                        {
                            filter: new ListAttributeValueBuilder()
                                .withUniverse(["enum_value_1", "enum_value_2"])
                                .withType("Enum")
                                .withFilterable(true)
                                .build()
                        }
                    ]
                };
                const headerFilterStore = new HeaderFiltersStore(props, headerFilterStoreInfo, null);
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = createContext<FilterAPIv2>(
                    headerFilterStore.context
                );
            });

            it("renders error message", () => {
                const { container } = render(
=======
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                    filterDispatcher: jest.fn(),
                    singleAttribute: new ListAttributeValueBuilder()
                        .withUniverse(["enum_value_1", "enum_value_2"])
                        .withType("Enum")
                        .withFilterable(true)
                        .build()
                } as FilterContextValue);
            });

            it("renders error message", () => {
                const filter = mount(
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
                    <DatagridDropdownFilter
                        {...commonProps}
                        auto={false}
                        multiSelect={false}
                        filterOptions={[
                            {
                                caption: dynamicValue<string>("wrong value"),
                                value: dynamicValue<string>("enum_value_3")
                            }
                        ]}
                    />
                );

<<<<<<< HEAD
                expect(container.querySelector(".alert")?.textContent).toBe("Invalid option value: 'enum_value_3'");
=======
                expect(filter.find(Alert).text()).toBe(
                    "Error: Invalid option (wrong value): option has invalid value and can't be used with attribute(s)"
                );
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            });
        });

        describe("with multiple invalid values", () => {
            beforeAll(() => {
<<<<<<< HEAD
                const props: HeaderFiltersStoreProps = {
                    filterList: [
                        {
                            filter: new ListAttributeValueBuilder()
                                .withUniverse(["enum_value_1", "enum_value_2"])
                                .withType("Enum")
                                .withFilterable(true)
                                .build()
                        },
                        {
                            filter: new ListAttributeValueBuilder()
                                .withUniverse([true, false])
                                .withType("Boolean")
                                .withFilterable(true)
                                .build()
                        }
                    ]
                };
                const headerFilterStore = new HeaderFiltersStore(props, headerFilterStoreInfo, null);
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = createContext<FilterAPIv2>(
                    headerFilterStore.context
                );
            });

            it("renders error message", () => {
                const { container } = render(
=======
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                    filterDispatcher: jest.fn(),
                    multipleAttributes: {
                        attribute1: new ListAttributeValueBuilder()
                            .withUniverse(["enum_value_1", "enum_value_2"])
                            .withType("Enum")
                            .withFilterable(true)
                            .build(),
                        attribute2: new ListAttributeValueBuilder()
                            .withUniverse([true, false])
                            .withType("Boolean")
                            .withFilterable(true)
                            .build()
                    }
                } as FilterContextValue);
            });

            it("renders error message", () => {
                const filter = mount(
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
                    <DatagridDropdownFilter
                        {...commonProps}
                        auto={false}
                        multiSelect={false}
                        filterOptions={[
                            {
                                caption: dynamicValue<string>("wrong enum value"),
                                value: dynamicValue<string>("enum_value_3")
                            },
                            {
                                caption: dynamicValue<string>("wrong boolean value"),
                                value: dynamicValue<string>("no")
                            }
                        ]}
                    />
                );

<<<<<<< HEAD
                expect(container.querySelector(".alert")?.textContent).toBe("Invalid option value: 'enum_value_3'");
=======
                expect(filter.find(Alert).text()).toBe(
                    "Error: Invalid option (wrong enum value): option has invalid value and can't be used with attribute(s)"
                );
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            });
        });
    });

    describe("with multiple instances", () => {
        beforeAll(() => {
<<<<<<< HEAD
            const props: HeaderFiltersStoreProps = {
                filterList: [
                    {
                        filter: new ListAttributeValueBuilder()
                            .withUniverse(["enum_value_1", "enum_value_2"])
                            .withType("Enum")
                            .withFilterable(true)
                            .withFormatter(
                                value => value,
                                () => console.log("Parsed")
                            )
                            .build()
                    }
                ]
            };
            const headerFilterStore = new HeaderFiltersStore(props, headerFilterStoreInfo, null);
            (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = createContext<FilterAPIv2>(
                headerFilterStore.context
            );
=======
            (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                filterDispatcher: jest.fn(),
                singleAttribute: new ListAttributeValueBuilder()
                    .withUniverse(["enum_value_1", "enum_value_2"])
                    .withType("Enum")
                    .withFilterable(true)
                    .withFormatter(
                        value => value,
                        () => console.log("Parsed")
                    )
                    .build()
            } as FilterContextValue);
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
        });

        it("renders with a unique id", () => {
            const { asFragment: fragment1 } = render(
                <DatagridDropdownFilter {...commonProps} auto multiSelect={false} filterOptions={[]} />
            );
            const { asFragment: fragment2 } = render(
                <DatagridDropdownFilter {...commonProps} auto multiSelect={false} filterOptions={[]} />
            );

            expect(fragment1().querySelector("input")?.getAttribute("aria-controls")).not.toBe(
                fragment2().querySelector("input")?.getAttribute("aria-controls")
            );
        });

        afterAll(() => {
<<<<<<< HEAD
            (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = undefined;
=======
            (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            delete (global as any)["com.mendix.widgets.web.UUID"];
        });
    });
});
