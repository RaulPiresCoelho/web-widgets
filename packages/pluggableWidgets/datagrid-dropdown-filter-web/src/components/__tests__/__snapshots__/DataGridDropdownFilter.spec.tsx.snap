// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Filter with single instance with single attribute DOM structure renders correctly 1`] = `
<DocumentFragment>
  <div
    class="dropdown-container filter-custom-class"
    data-focusindex="0"
  >
    <input
<<<<<<< HEAD
      aria-controls="Dropdown1-dropdown-list"
      aria-expanded="false"
      aria-haspopup="true"
      class="form-control dropdown-triggerer"
      placeholder=""
=======
      aria-controls="DropdownFilter1-dropdown-list"
      aria-expanded="false"
      aria-haspopup="true"
      class="form-control dropdown-triggerer"
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
      value=""
    />
  </div>
</DocumentFragment>
`;
